File: src/utils/date.ts
// Centralized date formatting utilities

/**
 * Format date as "D Mon" or "D Mon YYYY" (European style)
 */
export function formatDate(date: Date, options?: { includeYear?: boolean }): string {
  return date.toLocaleDateString('en-GB', {
    day: 'numeric',
    month: 'short',
    year: options?.includeYear ? 'numeric' : undefined,
  })
}

/**
 * Format date with full weekday and month names (for markdown headings)
 */
export function formatDateLong(date: Date): string {
  return date.toLocaleDateString('en-GB', {
    weekday: 'long',
    day: 'numeric',
    month: 'long',
    year: 'numeric',
  })
}

/**
 * Format time as HH:MM (24-hour)
 */
export function formatTime(date: Date): string {
  return date.toLocaleTimeString('en-GB', {
    hour: '2-digit',
    minute: '2-digit',
    hour12: false,
  })
}

/**
 * Format as relative time using Intl.RelativeTimeFormat
 * e.g., "2 days ago", "5 minutes ago", "just now"
 */
export function formatRelativeTime(date: Date): string {
  const now = Date.now()
  const diff = date.getTime() - now // negative for past
  const absDiff = Math.abs(diff)

  const rtf = new Intl.RelativeTimeFormat('en', { numeric: 'auto' })

  const minute = 60 * 1000
  const hour = 60 * minute
  const day = 24 * hour
  const week = 7 * day
  const month = 30 * day
  const year = 365 * day

  if (absDiff < minute) {
    return 'just now'
  } else if (absDiff < hour) {
    return rtf.format(Math.round(diff / minute), 'minute')
  } else if (absDiff < day) {
    return rtf.format(Math.round(diff / hour), 'hour')
  } else if (absDiff < week) {
    return rtf.format(Math.round(diff / day), 'day')
  } else if (absDiff < month) {
    return rtf.format(Math.round(diff / week), 'week')
  } else if (absDiff < year) {
    return rtf.format(Math.round(diff / month), 'month')
  } else {
    return rtf.format(Math.round(diff / year), 'year')
  }
}

/**
 * Get date group label for message grouping.
 * Returns "Today", "Yesterday", or formatted date.
 */
export function getDateGroupLabel(date: Date): string {
  const today = new Date()
  const yesterday = new Date(today)
  yesterday.setDate(yesterday.getDate() - 1)

  if (date.toDateString() === today.toDateString()) {
    return 'Today'
  }
  if (date.toDateString() === yesterday.toDateString()) {
    return 'Yesterday'
  }

  return formatDate(date, {
    includeYear: date.getFullYear() !== today.getFullYear(),
  })
}

File: internal/output/date.go
package output

import (
	"fmt"
	"math"
	"time"
)

type DateFormatter struct {
	now      func() time.Time
	location *time.Location
}

func NewDateFormatter(now func() time.Time, location *time.Location) DateFormatter {
	if now == nil {
		panic("output: nil clock")
	}
	if location == nil {
		panic("output: nil location")
	}
	return DateFormatter{now: now, location: location}
}

func (f DateFormatter) FormatDate(date time.Time, includeYear bool) string {
	date = date.In(f.location)
	if includeYear {
		return fmt.Sprintf("%d %s %d", date.Day(), date.Format("Jan"), date.Year())
	}
	return fmt.Sprintf("%d %s", date.Day(), date.Format("Jan"))
}

func (f DateFormatter) FormatDateLong(date time.Time) string {
	return date.In(f.location).Format("Monday, 2 January 2006")
}

func (f DateFormatter) FormatTime(date time.Time) string {
	return date.In(f.location).Format("15:04")
}

func (f DateFormatter) FormatRelativeTime(date time.Time) string {
	diffMillis := date.UnixMilli() - f.now().UnixMilli()
	absMillis := diffMillis
	if absMillis < 0 {
		absMillis = -absMillis
	}
	unitMillis := int64(time.Second / time.Millisecond)
	name := ""
	switch {
	case absMillis < int64(time.Minute/time.Millisecond):
		return "just now"
	case absMillis < int64(time.Hour/time.Millisecond):
		unitMillis, name = int64(time.Minute/time.Millisecond), "minute"
	case absMillis < int64((24*time.Hour)/time.Millisecond):
		unitMillis, name = int64(time.Hour/time.Millisecond), "hour"
	case absMillis < int64((7*24*time.Hour)/time.Millisecond):
		unitMillis, name = int64((24*time.Hour)/time.Millisecond), "day"
	case absMillis < int64((30*24*time.Hour)/time.Millisecond):
		unitMillis, name = int64((7*24*time.Hour)/time.Millisecond), "week"
	case absMillis < int64((365*24*time.Hour)/time.Millisecond):
		unitMillis, name = int64((30*24*time.Hour)/time.Millisecond), "month"
	default:
		unitMillis, name = int64((365*24*time.Hour)/time.Millisecond), "year"
	}
	value := int64(math.Floor(float64(diffMillis)/float64(unitMillis) + 0.5)) // JavaScript Math.round
	if name == "day" && value == -1 {
		return "yesterday"
	}
	if name == "day" && value == 1 {
		return "tomorrow"
	}
	if value == -1 && (name == "week" || name == "month" || name == "year") {
		return "last " + name
	}
	if value == 1 && (name == "week" || name == "month" || name == "year") {
		return "next " + name
	}
	label := name
	if value != -1 && value != 1 {
		label += "s"
	}
	if value < 0 {
		return fmt.Sprintf("%d %s ago", -value, label)
	}
	return fmt.Sprintf("in %d %s", value, label)
}

func (f DateFormatter) DateGroupLabel(date time.Time) string {
	now := f.now().In(f.location)
	date = date.In(f.location)
	today := time.Date(now.Year(), now.Month(), now.Day(), 0, 0, 0, 0, f.location)
	day := time.Date(date.Year(), date.Month(), date.Day(), 0, 0, 0, 0, f.location)
	if day.Equal(today) {
		return "Today"
	}
	if day.Equal(today.AddDate(0, 0, -1)) {
		return "Yesterday"
	}
	return f.FormatDate(date, date.Year() != now.Year())
}
